### DISCUSSION TITLE
Why not null-ls in this config?

### DISCUSSION THREAD

**Discussion by brxxlstxrs** [ayamir/nvimdots#500] [Q&A] [answered]
### Feature description

null-ls is an attempt to reduce the boilerplate required to set up general-purpose language servers and improve performance by removing the need for external processes.

### Additional information

maybe you use a substitute for this and I didn't notice, tell me?

**Comment by charliie-dev** [comment]
Hi, it would be better to ask this kind of questions in https://github.com/ayamir/nvimdots/discussions.

**Comment by charliie-dev** [comment, marked as answer]
In this config, we use [efm](https://github.com/creativenull/efmls-configs-nvim) to do the linting and formatting.

https://github.com/ayamir/nvimdots/blob/a6998aeef29f7a6674c891c84b89847c6808a3f5/lua/modules/plugins/completion.lua#L8

https://github.com/ayamir/nvimdots/blob/a6998aeef29f7a6674c891c84b89847c6808a3f5/lua/modules/configs/completion/lsp.lua#L117-L176

Then why `efm` instead of `null-ls` in the first place? That's b/c `efm` runs as daemons, and this reduce the startup time when firing up your nvim.

But https://github.com/creativenull/efmls-configs-nvim seems to been not updated in 4 months, maybe it's time to transfer to [null-ls](https://github.com/jose-elias-alvarez/null-ls.nvim)? 
For: 
1. Linters and formatters aren't really LSPs, it's been called only when saved. IMO, it doesn't need to run as daemon in the most case.
2. `Null-ls` contains wider amount of linters and formatters comparing with `efm` which is great for users to config new linters and formatters on their own.

@ayamir @Jint-lzxy How do u think?

**Reply by charliie-dev** [reply]
cc #494

**Comment by ayamir** [comment]
`efm` was configured as linters and formatters manager about this time last year. The reason why I use `efm` is `null-ls.nvim` just WIP and beta in that time, though it still claims it's in beta state now. We can switch to `null-ls.nvim` but it's a big change, maybe new version release?

**Reply by charliie-dev** [reply]
I use `null-ls` in my personal config for quite a while, and get no problem at all. In fact, we don't need `lua/modules/configs/completion/formatting.lua` if we set `null-ls` properly. Will draft a pr after I finished #495
